_posts/2020-12-04-release-1.12.0.md

```diff
@@ -0,0 +1,332 @@
+
+This release concluded the work started in Flink 1.9 on a [new data type system]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/table/types.html#data-types) for the Table API, with the exposure of aggregate functions (UDAFs) to the new type system. From Flink 1.12, UDAFs behave similarly to scalar and table functions, and support all data types.
+
+<hr>
+
+### PyFlink: Python DataStream API
+
+To expand the usability of PyFlink, this release introduces a first version of the Python DataStream API ([FLIP-130](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=158866298)) with support for stateless operations (e.g. Map, FlatMap, Filter, KeyBy). 
+
+```python
+from pyflink.common.typeinfo import Types
+from pyflink.datastream import MapFunction, StreamExecutionEnvironment
+
+class MyMapFunction(MapFunction):
+
+    def map(self, value):
+        return value + 1
+
+
+env = StreamExecutionEnvironment.get_execution_environment()
+data_stream = env.from_collection([1, 2, 3, 4, 5], type_info=Types.INT())
+mapped_stream = data_stream.map(MyMapFunction(), output_type=Types.INT())
+mapped_stream.print()
+env.execute("datastream job")
+```
+
+To give the Python DataStream API a try, you can [install PyFlink]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/python/installation.html#installation-of-pyflink) and check out [this tutorial]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/python/datastream_tutorial.html) that guides you through building a simple streaming application.
+
+<hr>
+
+### Other Improvements to PyFlink
+
+**PyFlink Jobs on Kubernetes ([FLINK-17480](https://issues.apache.org/jira/browse/FLINK-17480))**
+
+In addition to standalone and YARN deployments, PyFlink jobs can now also be deployed natively on Kubernetes. The [deployment documentation]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/deployment/resource-providers/native_kubernetes.html) has detailed instructions on how to start a _session_ or _application_ cluster on Kubernetes.
+
+**User-defined Aggregate Functions (UDAFs)**
+
+From Flink 1.12, you can define and register UDAFs in PyFlink ([FLIP-137](https://cwiki.apache.org/confluence/display/FLINK/FLIP-137%3A+Support+Pandas+UDAF+in+PyFlink?src=jira)). In contrast to a normal UDF, which doesn’t handle state and operates on a single row at a time, a [UDAF]({{ site.DOCS_BASE_URL }}flink-docs-release-1.12/dev/python/table-api-users-guide/udfs/python_udfs.html#aggregate-functions) is stateful and can be used to compute custom aggregations over multiple input rows. To benefit from vectorization, you can also use [Pandas UDAFs](https://ci.apache.org/projects/flink/flink-docs-release-1.12/dev/python/table-api-users-guide/udfs/vectorized_python_udfs.html#vectorized-aggregate-functions) ([FLIP-139](https://cwiki.apache.org/confluence/display/FLINK/FLIP-139%3A+General+Python+User-Defined+Aggregate+Function+Support+on+Table+API)) (up to 10x faster).
```

Comment:
"The FLIP and links for Pandas UDAF and general Python UDAF should be exchanged. Pandas UDAF is supported in FLIP-137 and general Python UDAF is supported in FLIP-139 and so we need to exchange them."